### PR TITLE
Several code improvements

### PR DESCRIPTION
**Pull request by frankiek3**
Several code improvements that should help speed up host of mod.

Reference #840

### CODE REVIEW

File: Modules/Utils.cs [hunk 1]
```diff
@@ -122,14 +122,13 @@ public static (string, Color) GetRoleTextHideAndSeek(RoleTypes oRole, CustomRole
...
+            if (p.Tasks == null || p.Role == null)
+                return false;
 
             var hasTasks = true;
-            if (p.Disconnected) hasTasks = false;
-            if (p.Role.IsImpostor)
+            if (p.Disconnected || p.Role.IsImpostor)
                 hasTasks = false; //タスクはCustomRoleを元に判定する
-            if (Options.CurrentGameMode == CustomGameMode.HideAndSeek)
+            else if (Options.CurrentGameMode == CustomGameMode.HideAndSeek)
```
Comment (soukunsandesu) [inline Modules/Utils.cs:131]: Impostor saves are counted in HideAndSeek

File: Modules/Utils.cs [hunk 2]
```diff
@@ -621,60 +627,91 @@ public static void NotifyRoles(bool isMeeting = false, PlayerControl SpecifySeer
...
+                                TargetMark += $"<color={GetRoleColorCode(CustomRoles.Lovers)}>♡</color>";
+                            }
 
-                        //ハートマークを付ける(相手に)
-                        if (seer.Is(CustomRoles.Lovers) && target.Is(CustomRoles.Lovers))
-                        {
-                            TargetMark += $"<color={GetRoleColorCode(CustomRoles.Lovers)}>♡</color>";
+                            foreach (var ExecutionerTarget in Main.ExecutionerTarget)
+                            {
+                                if (target.PlayerId == ExecutionerTarget.Value)
```
Comment (soukunsandesu) [inline Modules/Utils.cs:660]: Targets of other Executioners are visible

Comment (frankiek3) [reply Modules/Utils.cs:660]: The dead can see all targets, should only dead Executioners see their failed target?

File: Modules/Utils.cs (hunk 1 above)
Comment (frankiek3) [reply Modules/Utils.cs:131]: Fixed with 7b78cb36f81204ab7423765d3e9c128809b5d12d

**Comment by frankiek3** [comment]
Also fixed bug in ending screen 13b53cd0d9332b2737301cd6ab164d0094cd51ec

File: Modules/Utils.cs (hunk 1 above)
Comment (soukunsandesu) [reply Modules/Utils.cs:131]: No, don’t use “else”

**Comment by soukunsandesu** [comment]
Will the end screen change with the winner as Impostor? ??

File: Modules/Utils.cs (hunk 1 above)
Comment (frankiek3) [reply Modules/Utils.cs:131]: Revert back? 
```
...
            var hasTasks = true;
            if (p.Disconnected) hasTasks = false;
            if (p.Role.IsImpostor)
                hasTasks = false; //タスクはCustomRoleを元に判定する
            if (Options.CurrentGameMode == CustomGameMode.HideAndSeek)
            {
                if (p.IsDead) hasTasks = false;
                var hasRole = Main.AllPlayerCustomRoles.TryGetValue(p.PlayerId, out var role);
                if (hasRole)
                {
                    if (role is CustomRoles.HASFox or CustomRoles.HASTroll) hasTasks = false;
                }
            }
            else
...
```

Suggestions?

File: Modules/RPC.cs [hunk 3]
```diff
@@ -403,21 +409,18 @@ public static void SendExecutionerTarget(byte executionerId, byte targetId)
...
-            }
             if (AmongUsClient.Instance.AmHost)
             {
-                foreach (var imp in Impostors)
+                foreach (var p in PlayerControl.AllPlayerControls)
                 {
-                    imp.RpcSetRole(RoleTypes.GuardianAngel);
+                    if (p.PlayerId == winnerID)
+                    {
+                        p.RpcSetRole(RoleTypes.Imposter)
```
Comment (soukunsandesu) [inline Modules/RPC.cs:418]: typo

File: Modules/Utils.cs (hunk 1 above)
Comment (soukunsandesu) [reply Modules/Utils.cs:131]: Yes, please revert.

**Comment by frankiek3** [comment]
Jester end screen currently (2.1.1 known) displays imposters winning with no imposters listed. I found the bugged code by mistake and made a mess so I thought I'd clean it up.
End screen should be imposter win with the one solo winner shown after fix.

File: Modules/RPC.cs [hunk 4]
```diff
@@ -357,12 +357,18 @@ public static void SetCustomRole(byte targetId, CustomRoles role)
...
-            if (role == CustomRoles.Sniper) Sniper.Add(targetId);
+            if (role == CustomRoles.FireWorks)
+            {
+                FireWorks.Add(targetId);
+            }
+            else if (role == CustomRoles.Sniper)
+            {
+                Sniper.Add(targetId);
+            }
             HudManager.Instance.SetHudActive(true);
```
Comment (yurinakira) [inline Modules/RPC.cs:372]: setCustomRoleは高い頻度で呼ばれる処理でもないので、"高速化"とは違うかと.

**Comment by soukunsandesu** [comment]
Are you do test?

File: Modules/RPC.cs (hunk 3 above)
Comment (frankiek3) [reply Modules/RPC.cs]: Typo fixed 8b0e391dbe1071e408c1cb639ba93682345cb85b

File: Modules/RPC.cs (hunk 4 above)
Comment (frankiek3) [reply Modules/RPC.cs:372]: Correct, I guess not :(

**Comment by frankiek3** [comment]
I'm not setup for Visual Studio on Windows 10, but can have that installed if needed.

**Comment by yurinakira** [comment]
Please do not do anything different from the Pull requests Title.
If you want to do something different, please do a separate Pull Requests.

File: Modules/Utils.cs (hunk 1 above)
Comment (frankiek3) [reply Modules/Utils.cs]: Reference
746411adfa7db7a19da61af3eb5b347ac1d8b2d1

File: Modules/RPC.cs [hunk 5]
```diff
@@ -403,21 +409,18 @@ public static void SendExecutionerTarget(byte executionerId, byte targetId)
...
-                    imp.RpcSetRole(RoleTypes.GuardianAngel);
+                    if (p.PlayerId == winnerID)
+                    {
+                        p.RpcSetRole(RoleTypes.Impostor)
+                    }
+                    else if(p.Data.Role.IsImpostor)
+                    {
+                        p.RpcSetRole(RoleTypes.GuardianAngel);
+                    }
                 }
```
Comment (yurinakira) [inline Modules/RPC.cs:424]: ちょっと微妙かな

File: Modules/RPC.cs [hunk 6]
```diff
@@ -357,12 +357,18 @@ public static void SetCustomRole(byte targetId, CustomRoles role)
             {
                 Main.AllPlayerCustomRoles[targetId] = role;
             }
-            else if (role >= CustomRoles.NoSubRoleAssigned)   //500:NoSubRole 501~:SubRole
+            else //if (role >= CustomRoles.NoSubRoleAssigned)   //500:NoSubRole 501~:SubRole
```
[the remark this answers is not in the document]
Comment (frankiek3) [reply Modules/RPC.cs]: Uncommented now 
ff6a9b74401ff713d4ca4a2b822bee2f72267f85

File: Modules/RPC.cs (hunk 5 above)
Comment (frankiek3) [reply Modules/RPC.cs:421]: Removed bug fix and reverted Winner code back
Will fix bug in separate pull request 
ff6a9b74401ff713d4ca4a2b822bee2f72267f85

File: Modules/RPC.cs [hunk 7]
```diff
@@ -403,25 +409,20 @@ public static void SendExecutionerTarget(byte executionerId, byte targetId)
         }
         public static void CustomWinTrigger(byte winnerID)
         {
-            List<PlayerControl> Impostors = new();
+            if (!AmongUsClient.Instance.AmHost) return;
+
+            PlayerControl Winner = null;
```
Comment (soukunsandesu) [inline Modules/RPC.cs:414]: Not require

Comment (frankiek3) [reply Modules/RPC.cs:414]: Remove unused variable Winner and associated if statement?

Comment (soukunsandesu) [reply Modules/RPC.cs:414]: If statement require

**Comment by soukunsandesu** [comment]
If you can build, I will approve this pull request.

**Comment by frankiek3** [comment]
I'm was getting an error with the same variable on 568 and 585. Now it builds.

https://github.com/tukasa0001/TownOfHost/blob/9bebcaf62ce2c17b863d0e6c489d2c2b88fe8d09/Modules/Utils.cs#L568-L586

**Review by soukunsandesu — APPROVED** [review]

**Comment by soukunsandesu** [comment]
もう一人分ほしいです

**Comment by yurinakira** [comment]
ぱっと見問題なさそうだが次リリース後にマージしたい

**Comment by yurinakira** [comment]
このPR古くなっちゃったんで取り下げませんか...?

**Pull request closed without merging**